File: vcl/java/source/java/VCLScreen.cxx
/*************************************************************************
 *
 *  $RCSfile$
 *
 *  $Revision$
 *
 *  last change: $Author$ $Date$
 *
 *  The Contents of this file are made available subject to the terms of
 *  either of the following licenses
 *
 *         - GNU General Public License Version 2.1
 *
 *  Patrick Luby, June 2003
 *
 *  GNU General Public License Version 2.1
 *  =============================================
 *  Copyright 2003 Planamesa Inc.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License version 2.1, as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *  MA  02111-1307  USA
 *
 ************************************************************************/

#define _SV_COM_SUN_STAR_VCL_VCLSCREEN_CXX

#ifndef _SV_COM_SUN_STAR_VCL_VCLSCREEN_HXX
#include <com/sun/star/vcl/VCLScreen.hxx>
#endif

using namespace vcl;

// ============================================================================

static jobject JNICALL Java_com_sun_star_vcl_VCLScreen_getScreenInsets( JNIEnv *pEnv, jobject object, jobject _par0 )
{
	jobject out = NULL;

	if ( _par0 )
	{
		jclass tempClass = pEnv->FindClass( "apple/awt/CGraphicsDevice" );
		if ( tempClass && pEnv->IsInstanceOf( _par0, tempClass ) )
		{
			static jmethodID mIDGetScreenInsets = NULL;
			if ( !mIDGetScreenInsets )
            {
				char *cSignature = "()Ljava/awt/Insets;";
				mIDGetScreenInsets = pEnv->GetMethodID( tempClass, "getScreenInsets", cSignature );
			}
			OSL_ENSURE( mIDGetScreenInsets, "Unknown method id!" );
			if ( mIDGetScreenInsets )
				out = pEnv->CallObjectMethod( _par0, mIDGetScreenInsets );
		}
	}

	return out;
}

// ============================================================================

jclass com_sun_star_vcl_VCLScreen::theClass = NULL;

// ----------------------------------------------------------------------------

jclass com_sun_star_vcl_VCLScreen::getMyClass()
{
	if ( !theClass )
	{
		VCLThreadAttach t;
		if ( !t.pEnv ) return (jclass)NULL;
		jclass tempClass = t.pEnv->FindClass( "com/sun/star/vcl/VCLScreen" );
		OSL_ENSURE( tempClass, "Java : FindClass not found!" );

		if ( tempClass )
		{
			// Register the native methods for our class
			JNINativeMethod aMethod;
			aMethod.name = "getScreenInsets";
			aMethod.signature = "(Ljava/awt/GraphicsDevice;)Ljava/awt/Insets;";
			aMethod.fnPtr = (void *)Java_com_sun_star_vcl_VCLScreen_getScreenInsets;
			t.pEnv->RegisterNatives( tempClass, &aMethod, 1 );
		}

		theClass = (jclass)t.pEnv->NewGlobalRef( tempClass );
	}
	return theClass;
}

// ----------------------------------------------------------------------------

SalColor com_sun_star_vcl_VCLScreen::getControlColor()
{
	static jmethodID mID = NULL;
	SalColor out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getControlColor", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (SalColor)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

unsigned int com_sun_star_vcl_VCLScreen::getDefaultScreenNumber()
{
	static jmethodID mID = NULL;
	unsigned int out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getDefaultScreenNumber", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (unsigned int)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

const Rectangle com_sun_star_vcl_VCLScreen::getScreenBounds( long _par0, long _par1, long _par2, long _par3, sal_Bool _par4, sal_Bool _par5 )
{
	static jmethodID mID = NULL;
	static jfieldID fIDX = NULL;
	static jfieldID fIDY = NULL;
	static jfieldID fIDWidth = NULL;
	static jfieldID fIDHeight = NULL;
	Rectangle out( Point( 0, 0 ), Size( 0, 0 ) );
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "(IIIIZZ)Ljava/awt/Rectangle;";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getScreenBounds", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jvalue args[6];
			args[0].i = jint( _par0 );
			args[1].i = jint( _par1 );
			args[2].i = jint( _par2 );
			args[3].i = jint( _par3 );
			args[4].z = jboolean( _par4 );
			args[5].z = jboolean( _par5 );
			jobject tempObj = t.pEnv->CallStaticObjectMethodA( getMyClass(), mID, args );
			if ( tempObj )
			{
				jclass tempObjClass = t.pEnv->GetObjectClass( tempObj );
				if ( !fIDX )
				{
					char *cSignature = "I";
					fIDX = t.pEnv->GetFieldID( tempObjClass, "x", cSignature );
				}
				OSL_ENSURE( fIDX, "Unknown field id!" );
				if ( !fIDY )
				{
					char *cSignature = "I";
					fIDY = t.pEnv->GetFieldID( tempObjClass, "y", cSignature );
				}
				OSL_ENSURE( fIDY, "Unknown field id!" );
				if ( !fIDWidth )
				{
					char *cSignature = "I";
					fIDWidth = t.pEnv->GetFieldID( tempObjClass, "width", cSignature );
				}
				OSL_ENSURE( fIDWidth, "Unknown field id!" );
				if ( !fIDHeight )
				{
					char *cSignature = "I";
					fIDHeight = t.pEnv->GetFieldID( tempObjClass, "height", cSignature );
				}
				OSL_ENSURE( fIDHeight, "Unknown field id!" );
				if ( fIDX && fIDY && fIDWidth && fIDHeight )
				{
					Point aPoint( (long)t.pEnv->GetIntField( tempObj, fIDX ), (long)t.pEnv->GetIntField( tempObj, fIDY ) );
					Size aSize( (long)t.pEnv->GetIntField( tempObj, fIDWidth ), (long)t.pEnv->GetIntField( tempObj, fIDHeight ) );
					out = Rectangle( aPoint, aSize );
				}
			}
		}
	}
	return out;
}

// ----------------------------------------------------------------------------

const Rectangle com_sun_star_vcl_VCLScreen::getScreenBounds( unsigned int _par0, sal_Bool _par1 )
{
	static jmethodID mID = NULL;
	static jfieldID fIDX = NULL;
	static jfieldID fIDY = NULL;
	static jfieldID fIDWidth = NULL;
	static jfieldID fIDHeight = NULL;
	Rectangle out( Point( 0, 0 ), Size( 0, 0 ) );
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "(IZ)Ljava/awt/Rectangle;";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getScreenBounds", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jvalue args[2];
			args[0].i = jint( _par0 );
			args[1].z = jboolean( _par1 );
			jobject tempObj = t.pEnv->CallStaticObjectMethodA( getMyClass(), mID, args );
			if ( tempObj )
			{
				jclass tempObjClass = t.pEnv->GetObjectClass( tempObj );
				if ( !fIDX )
				{
					char *cSignature = "I";
					fIDX = t.pEnv->GetFieldID( tempObjClass, "x", cSignature );
				}
				OSL_ENSURE( fIDX, "Unknown field id!" );
				if ( !fIDY )
				{
					char *cSignature = "I";
					fIDY = t.pEnv->GetFieldID( tempObjClass, "y", cSignature );
				}
				OSL_ENSURE( fIDY, "Unknown field id!" );
				if ( !fIDWidth )
				{
					char *cSignature = "I";
					fIDWidth = t.pEnv->GetFieldID( tempObjClass, "width", cSignature );
				}
				OSL_ENSURE( fIDWidth, "Unknown field id!" );
				if ( !fIDHeight )
				{
					char *cSignature = "I";
					fIDHeight = t.pEnv->GetFieldID( tempObjClass, "height", cSignature );
				}
				OSL_ENSURE( fIDHeight, "Unknown field id!" );
				if ( fIDX && fIDY && fIDWidth && fIDHeight )
				{
					Point aPoint( (long)t.pEnv->GetIntField( tempObj, fIDX ), (long)t.pEnv->GetIntField( tempObj, fIDY ) );
					Size aSize( (long)t.pEnv->GetIntField( tempObj, fIDWidth ), (long)t.pEnv->GetIntField( tempObj, fIDHeight ) );
					out = Rectangle( aPoint, aSize );
				}
			}
		}
	}
	return out;
}

// ----------------------------------------------------------------------------

unsigned int com_sun_star_vcl_VCLScreen::getScreenCount()
{
	static jmethodID mID = NULL;
	unsigned int out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getScreenCount", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (unsigned int)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

SalColor com_sun_star_vcl_VCLScreen::getTextHighlightColor()
{
	static jmethodID mID = NULL;
	SalColor out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getTextHighlightColor", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (SalColor)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

SalColor com_sun_star_vcl_VCLScreen::getTextHighlightTextColor()
{
	static jmethodID mID = NULL;
	SalColor out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getTextHighlightTextColor", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (SalColor)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

SalColor com_sun_star_vcl_VCLScreen::getTextTextColor()
{
	static jmethodID mID = NULL;
	SalColor out = 0;
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()I";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getTextTextColor", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
			out = (SalColor)t.pEnv->CallStaticIntMethod( getMyClass(), mID );
	}
	return out;
}

// ----------------------------------------------------------------------------

const Rectangle com_sun_star_vcl_VCLScreen::getVirtualScreenBounds()
{
	static jmethodID mID = NULL;
	static jfieldID fIDX = NULL;
	static jfieldID fIDY = NULL;
	static jfieldID fIDWidth = NULL;
	static jfieldID fIDHeight = NULL;
	Rectangle out( Point( 0, 0 ), Size( 0, 0 ) );
	VCLThreadAttach t;
	if ( t.pEnv )
	{
		if ( !mID )
		{
			char *cSignature = "()Ljava/awt/Rectangle;";
			mID = t.pEnv->GetStaticMethodID( getMyClass(), "getVirtualScreenBounds", cSignature );
		}
		OSL_ENSURE( mID, "Unknown method id!" );
		if ( mID )
		{
			jobject tempObj = t.pEnv->CallStaticObjectMethod( getMyClass(), mID );
			if ( tempObj )
			{
				jclass tempObjClass = t.pEnv->GetObjectClass( tempObj );
				if ( !fIDX )
				{
					char *cSignature = "I";
					fIDX = t.pEnv->GetFieldID( tempObjClass, "x", cSignature );
				}
				OSL_ENSURE( fIDX, "Unknown field id!" );
				if ( !fIDY )
				{
					char *cSignature = "I";
					fIDY = t.pEnv->GetFieldID( tempObjClass, "y", cSignature );
				}
				OSL_ENSURE( fIDY, "Unknown field id!" );
				if ( !fIDWidth )
				{
					char *cSignature = "I";
					fIDWidth = t.pEnv->GetFieldID( tempObjClass, "width", cSignature );
				}
				OSL_ENSURE( fIDWidth, "Unknown field id!" );
				if ( !fIDHeight )
				{
					char *cSignature = "I";
					fIDHeight = t.pEnv->GetFieldID( tempObjClass, "height", cSignature );
				}
				OSL_ENSURE( fIDHeight, "Unknown field id!" );
				if ( fIDX && fIDY && fIDWidth && fIDHeight )
				{
					Point aPoint( (long)t.pEnv->GetIntField( tempObj, fIDX ), (long)t.pEnv->GetIntField( tempObj, fIDY ) );
					Size aSize( (long)t.pEnv->GetIntField( tempObj, fIDWidth ), (long)t.pEnv->GetIntField( tempObj, fIDHeight ) );
					out = Rectangle( aPoint, aSize );
				}
			}
		}
	}
	return out;
}

File: vcl/java/com/sun/star/vcl/VCLScreen.java
/*************************************************************************
 *
 *  $RCSfile$
 *
 *  $Revision$
 *
 *  last change: $Author$ $Date$
 *
 *  The Contents of this file are made available subject to the terms of
 *  either of the following licenses
 *
 *         - GNU General Public License Version 2.1
 *
 *  Patrick Luby, June 2003
 *
 *  GNU General Public License Version 2.1
 *  =============================================
 *  Copyright 2003 Planamesa Inc.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public
 *  License version 2.1, as published by the Free Software Foundation.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *  MA  02111-1307  USA
 *
 ************************************************************************/

package com.sun.star.vcl;

import java.awt.Dimension;
import java.awt.Frame;
import java.awt.Graphics2D;
import java.awt.GraphicsConfiguration;
import java.awt.GraphicsDevice;
import java.awt.GraphicsEnvironment;
import java.awt.Insets;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.SystemColor;
import java.awt.Toolkit;
import java.awt.Window;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

/** 
 * The Java class that implements the convenience methods for accessing
 * screen information.
 * <p>
 * @version 	$Revision$ $Date$
 * @author 	    $Author$
 */
public final class VCLScreen {

	/**
	 * The minimum screen resolution.
	 */
	public final static int MIN_SCREEN_RESOLUTION = 96;

	/**
	 * The default screen bounds.
	 */
	private static Rectangle defaultScreenBounds = new Rectangle(0, 0, 800, 600);

	/**
	 * The cached frame insets.
	 */
	private static Insets frameInsets = null;

	/**
	 * The cached minimum frame size.
	 */
	private static Dimension minimumFrameSize = null;

	/**
	 * Cached buffered image.
	 */
	private static BufferedImage image = new BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB_PRE);

	/**
	 * Initialize screen size and frame insets.
	 */
	static {

		Frame f = new Frame();
		f.setSize(1, 1);
		f.addNotify();
		frameInsets = f.getInsets();
		minimumFrameSize = f.getSize();
		f.removeNotify();
		f.dispose();

	}


	/**
	 * Clears, refreshes, and returns the JVM's cached display list.
	 *
	 * @return the cached display list
	 */
	static GraphicsDevice[] clearCachedDisplays() {

		// Fix bug 3416 by forcing Java to refresh its screen list
		GraphicsDevice[] gd = null;

		GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
		if (ge != null) {
			synchronized (ge) {
				// Clear the existing JVM cache
				clearCachedDisplays0(ge);

				gd = ge.getScreenDevices();

				// On some machines, Java shows two monitors for every mirrored
				// display so eliminate those duplicate monitors
				if (gd != null && gd.length > 1) {
					Rectangle lastBounds = gd[0].getDefaultConfiguration().getBounds();
					ArrayList displays = new ArrayList(gd.length);
					displays.add(gd[0]);
					for (int i = 1; i < gd.length; i++) {
						Rectangle bounds = gd[i].getDefaultConfiguration().getBounds();
						if (bounds.width > 0 && bounds.height > 0 && !bounds.equals(lastBounds)) {
							displays.add(gd[i]);
							lastBounds = bounds;
						}
					}

					if (gd.length != displays.size()) {
						GraphicsDevice[] displayArray = new GraphicsDevice[displays.size()];
						displays.toArray(displayArray);
						gd = displayArray;
					}
				}
			}
		}

		return gd;

	}

	/**
	 * Clears the JVM's cached display list.
	 *
	 * @param ge the graphics environment
	 */
	public static native void clearCachedDisplays0(GraphicsEnvironment ge);

	/**
	 * Returns the <code>System.control</code>.
	 *
	 * @return the <code>System.control</code>
	 */
	public static int getControlColor() {

		int color = 0x00000000;

		Graphics2D g = image.createGraphics();
		if (g != null) {
			try {
				g.setColor(SystemColor.control);
				g.fillRect(0, 0, 1, 1);
				color = image.getRGB(0, 0);
			}
			catch (Throwable t) {
				t.printStackTrace();
			}
			g.dispose();
		}

		return color;

	}

	/**
	 * Return the default screen number.
	 *
	 * @return the default screen number
	 */
	public static int getDefaultScreenNumber() {

		GraphicsDevice[] gd = clearCachedDisplays();
		GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
		if (gd != null && ge != null) {
			GraphicsDevice dsd = ge.getDefaultScreenDevice();
			if (dsd != null && gd != null) {
				for (int i = 0; i < gd.length; i++) {
					if (dsd == gd[i])
						return i;
				}
			}
		}

		return 0;

	}

	/**
	 * Gets the <code>Frame</code> insets.
	 *
	 * @return the <code>Frame</code> insets
	 */
	public static Insets getFrameInsets() {

		return frameInsets;

	}

	/**
	 * Gets the minimum <code>Frame</code> size.
	 *
	 * @return the minimum <code>Frame</code> size
	 */
	public static Dimension getMinimumFrameSize() {

		return minimumFrameSize;

	}

	/**
	 * Returns the screen bounds that is the closest match for the specified
	 * bounds.
	 *
	 * @param x the x coordinate 
	 * @param y the y coordinate 
	 * @param width the width
	 * @param height the height
	 * @param fullScreenMode if <code>true</code> ignore screen insets
	 *  otherwise include screen insets in the screen bounds
	 * @return the screen bounds that is the closest match for the specified
	 *  coordinates
	 */
	public static Rectangle getScreenBounds(int x, int y, int width, int height, boolean fullScreenMode ) {

		GraphicsDevice[] gd = clearCachedDisplays();
		if (gd != null) {
			// Fix bug 2671 by setting width and height greater than 0
			if (width <= 0)
				width = 1;
			if (height <= 0)
				height = 1;

			// Iterate through the screens and find the screen that the
			// point is inside of
			for (int i = 0; i < gd.length; i++) {
				try {
					// Test if the point is inside the screen
					Rectangle r = gd[i].getDefaultConfiguration().getBounds();
					if (r.contains(x, y)) {
						if (!fullScreenMode) {
							Insets insets = VCLScreen.getScreenInsets(gd[i]);
							if (insets != null)
								r = new Rectangle(r.x + insets.left, r.y + insets.top, r.width - insets.left - insets.right, r.height - insets.top - insets.bottom);
						}
						return r;
					}
				}
				catch (Throwable t) {}
			}

			// Iterate through the screens and find the closest screen
			long closestArea = Long.MAX_VALUE;
			Rectangle closestBounds = null;
			for (int i = 0; i < gd.length; i++) {
				try {
					Rectangle r = gd[i].getDefaultConfiguration().getBounds();
					if (!fullScreenMode) {
						Insets insets = VCLScreen.getScreenInsets(gd[i]);
						if (insets != null)
							r = new Rectangle(r.x + insets.left, r.y + insets.top, r.width - insets.left - insets.right, r.height - insets.top - insets.bottom);
					}

					// Test the closeness of the point to the center of the
					// screen
					long area = Math.abs((r.x + (r.width / 2) - x) * (r.y + (r.height / 2) - y));
					if (closestArea > area) {
						closestArea = area;
						closestBounds = r;
					}
				}
				catch (Throwable t) {}
			}

			if (closestBounds == null || closestBounds.isEmpty())
				return VCLScreen.defaultScreenBounds;
			else
				return closestBounds;
		}

		return VCLScreen.defaultScreenBounds;

	}

	/**
	 * Returns the screen bounds for the specified screen number.
	 *
	 * @param n the screen number
	 * @param b <code>true</code> to obtain only the displayable screen area
	 *  and <code>false</code> to obtain the entire screen area
	 * @return the screen bounds or <code>null</code> if the specified
	 *  screen number does not exist
	 */
	public static Rectangle getScreenBounds(int n, boolean b) {

		GraphicsDevice[] gd = clearCachedDisplays();
		if (gd != null && n >= 0) {
			try {
				Rectangle r = gd[n].getDefaultConfiguration().getBounds();
				if (!r.isEmpty()) {
					if (b) {
						Insets insets = VCLScreen.getScreenInsets(gd[n]);
						if (insets != null)
							r = new Rectangle(r.x + insets.left, r.y + insets.top, r.width - insets.left - insets.right, r.height - insets.top - insets.bottom);
					}

					return r;
				}
			}
			catch (Throwable t) {}
		}

		return null;

	}

	/**
	 * Return the number of screens.
	 *
	 * @return the number of screens
	 */
	public static int getScreenCount() {

		GraphicsDevice[] gd = clearCachedDisplays();
		if (gd != null) {
			if (gd != null && gd.length > 1)
				return gd.length;
		}

		return 1;

	}

	/**
	 * Returns the insets for the specified <code>GraphicsDevice</code>.
	 *
	 * @param gd the <code>GraphicsDevice</code>
	 * @return the insets for the specified <code>GraphicsDevice</code>
	 */
	public static native Insets getScreenInsets(GraphicsDevice gd);

	/**
	 * Returns the <code>System.textHighlight</code>.
	 *
	 * @return the <code>System.textHighlight</code>
	 */
	public static int getTextHighlightColor() {

		int color = 0x00000000;

		Graphics2D g = image.createGraphics();
		if (g != null) {
			try {
				g.setColor(SystemColor.textHighlight);
				g.fillRect(0, 0, 1, 1);
				color = image.getRGB(0, 0);
			}
			catch (Throwable t) {
				t.printStackTrace();
			}
			g.dispose();
		}

		return color;

	}

	/**
	 * Returns the <code>System.textHighlightText</code>.
	 *
	 * @return the <code>System.textHighlightText</code>
	 */
	public static int getTextHighlightTextColor() {

		int color = 0x00000000;

		Graphics2D g = image.createGraphics();
		if (g != null) {
			try {
				g.setColor(SystemColor.textHighlightText);
				g.fillRect(0, 0, 1, 1);
				color = image.getRGB(0, 0);
			}
			catch (Throwable t) {
				t.printStackTrace();
			}
			g.dispose();
		}

		return color;

	}

	/**
	 * Returns the <code>System.textText</code>.
	 *
	 * @return the <code>System.textText</code>
	 */
	public static int getTextTextColor() {

		int color = 0x00000000;

		Graphics2D g = image.createGraphics();
		if (g != null) {
			try {
				g.setColor(SystemColor.textText);
				g.fillRect(0, 0, 1, 1);
				color = image.getRGB(0, 0);
			}
			catch (Throwable t) {
				t.printStackTrace();
			}
			g.dispose();
		}

		return color;

	}

}
